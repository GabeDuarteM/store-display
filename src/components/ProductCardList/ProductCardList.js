import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

import ProductCard from '../ProductCard'
import Button from '../Button'

export const styles = {
  productCardList: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '15px 0',
    margin: '0 47px',
    alignItems: 'center',
    justifyContent: 'center',
  },
  productCard: {
    flex: 'none',
    margin: 15,
  },
  loadMoreButtonRoot: {
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  loadMoreButton: {
    margin: 25,
  },
}

const ProductCardList = ({
  className,
  classes,
  products,
  toggleIsWishlisted,
  handleClickLoadMore,
  isLastPage,
}) => (
  <div className={cn(classes.productCardList, className)}>
    {products.map(product => (
      <ProductCard
        key={product.id}
        className={classes.productCard}
        toggleIsWishlisted={toggleIsWishlisted}
        {...product}
      />
    ))}
    <div className={classes.loadMoreButtonRoot}>
      {isLastPage === false && (
        <Button
          className={classes.loadMoreButton}
          onClick={handleClickLoadMore}
        >
          Load more
        </Button>
      )}
    </div>
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** Array of products to list */
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      imgUrl: PropTypes.string.isRequired,
      gender: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
      price: PropTypes.number.isRequired,
      isNew: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      brand: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /**
   * Function to add an item to a wishlist
   * @param {id} string The id of the item to toggle the wishlist state
   */
  toggleIsWishlisted: PropTypes.func.isRequired,
  /** Gets called when the user clicks on the Load more button */
  handleClickLoadMore: PropTypes.func.isRequired,
  /** Property that defines if there are more products to load or not  */
  isLastPage: PropTypes.bool,
  /** Apply that className to the root component */
  className: PropTypes.string,
}

export const defaultProps = {
  isLastPage: null,
  className: null,
}

ProductCardList.propTypes = propTypes
ProductCardList.defaultProps = defaultProps

export default injectSheet(styles)(ProductCardList)
