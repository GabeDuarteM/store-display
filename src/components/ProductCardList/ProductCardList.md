```js
<ProductCardList
  products={[
    {
      id: '5a5180507dceaccd20a314e0',
      imgUrl: '//images.arcteryx.com/F17/450x625/Cerium-LT-Jacket-Rooibos.png',
      gender: 'woman',
      rating: 9,
      price: 96.44,
      isNew: true,
      name: 'Aguirrewatson Ritakirkland',
      brand: 'Mazine',
    },
    {
      id: '5a518050fbb2da0a6d228cf9',
      imgUrl:
        '//cdn.shopify.com/s/files/1/1104/4168/products/Allbirds_W_Wool_Runner_Kotare_GREY_ANGLE_900x900.png?v=1514576081',
      gender: 'man',
      rating: 8,
      price: 105.7,
      isNew: false,
      name: 'Arlineschneider Loriemartinez',
      brand: 'NIKE',
    },
    {
      id: '5a51804f840a30ab47a631a2',
      imgUrl: '//www.kicksusa.com/media/wysiwyg/brands/nike/nike-men.png',
      gender: 'woman',
      rating: 10,
      price: 27.59,
      isNew: true,
      name: 'Candacechase Blackturner',
      brand: 'NIKE',
    },
    {
      id: '5a518050bbbe951b1e561fed',
      imgUrl:
        'http://shop.autismspeaks.org/Shared/images/Product/AUT/AT17012.png',
      gender: 'woman',
      rating: 4,
      price: 235.1,
      isNew: false,
      name: 'Burkemejia Cochranbray',
      brand: 'Urban Classics',
    },
    {
      id: '5a51805085f01f4226367dfd',
      imgUrl:
        'http://cdn.shopify.com/s/files/1/1378/9837/products/lets-love-tshirt_grande.png?v=1509580185',
      gender: 'man',
      rating: 2,
      price: 257.36,
      isNew: true,
      name: 'Fitzgeraldwallace Agnesmarks',
      brand: 'Urban Classics',
    },
    {
      id: '5a518050b6fb69f117f82d7e',
      imgUrl:
        '//cdn.shopify.com/s/files/1/1734/3501/products/CODWW2MLTRYSHRT-OLV-S_1024x1024.png?v=1507908119',
      gender: 'man',
      rating: 3,
      price: 270.8,
      isNew: false,
      name: 'Celiamoss Dominiquerosales',
      brand: 'Urban Classics',
    },
    {
      id: '5a518050864901ecde5be490',
      imgUrl:
        'https://www.swazi.co.nz/wp-content/uploads/2014/11/Micro-Pants-Olive-PMP.png',
      gender: 'woman',
      rating: 9,
      price: 119.6,
      isNew: false,
      name: 'Davidsonblackwell Fowlerbenson',
      brand: 'NIKE',
    },
    {
      id: '5a518050e421e659c6e24c46',
      imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1175/1060f.png',
      gender: 'man',
      rating: 10,
      price: 282.1,
      isNew: true,
      name: 'Perkinsjames Clinegalloway',
      brand: 'VERO MODA',
    },
    {
      id: '5a51805044d1030570bcab76',
      imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1025/1436f.png',
      gender: 'woman',
      rating: 4,
      price: 296.55,
      isNew: false,
      name: 'Henriettagillespie Dellavargas',
      brand: 'NIKE',
    },
  ]}
  handleClickLoadMore={() => {}}
  toggleIsWishlisted={() => {}}
  isLastPage={false}
/>
```
