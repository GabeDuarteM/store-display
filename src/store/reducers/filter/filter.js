import {
  FILTER__CHANGE_GENDER,
  FILTER__CHANGE_SORT,
  FILTER__LOAD_MORE_PRODUCTS,
} from './filterActions'

const returnInitialState = () => ({
  gender: 'woman',
  sort: 'price-asc',
  productListLength: 9,
})

const filter = (state = returnInitialState(), action = {}) => {
  switch (action.type) {
    case FILTER__CHANGE_GENDER:
      return {
        ...state,
        gender: action.payload,
        productListLength: 9,
      }
    case FILTER__CHANGE_SORT:
      return {
        ...state,
        sort: action.payload,
      }
    case FILTER__LOAD_MORE_PRODUCTS:
      return {
        ...state,
        productListLength: state.productListLength + 9,
      }
    default:
      return state
  }
}

export default filter
