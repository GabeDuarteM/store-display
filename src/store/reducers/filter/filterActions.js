export const FILTER__CHANGE_GENDER = 'FILTER/CHANGE_GENDER'
export const FILTER__CHANGE_SORT = 'FILTER/CHANGE_SORT'
export const FILTER__LOAD_MORE_PRODUCTS = 'FILTER/LOAD_MORE_PRODUCTS'

export const filterChangeGender = gender => ({
  type: FILTER__CHANGE_GENDER,
  payload: gender,
})

export const filterChangeSort = sort => ({
  type: FILTER__CHANGE_SORT,
  payload: sort,
})

export const filterLoadMoreProducts = () => ({
  type: FILTER__LOAD_MORE_PRODUCTS,
})
