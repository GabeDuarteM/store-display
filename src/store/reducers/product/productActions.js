import productListData from '../../../../data/products'

export const PRODUCT__SET_LIST = 'PRODUCT/SET_LIST'
export const PRODUCT__REFRESH_LIST = 'PRODUCT/REFRESH_LIST'
export const PRODUCT__SET_WISHLIST_PROP = 'PRODUCT/SET_WISHLIST_PROP'

const sortAsc = (itemA, itemB) => itemA - itemB
const sortDesc = (itemA, itemB) => itemB - itemA

const sortPrice = (sortType, itemA, itemB) => {
  const sortFunc = sortType === 'asc' ? sortAsc : sortDesc
  return sortFunc(itemA.price, itemB.price)
}

const sortRating = (sortType, itemA, itemB) => {
  const sortFunc = sortType === 'asc' ? sortAsc : sortDesc
  return sortFunc(itemA.rating, itemB.rating)
}

const sortProducts = sort => {
  const [prop, sortType] = sort.split('-')
  let sortFunc
  switch (prop) {
    case 'price':
      sortFunc = sortPrice
      break
    case 'rating':
      sortFunc = sortRating
      break
    default:
      throw new Error(`Invalid sort type: ${sort}`)
  }
  return sortFunc.bind(null, sortType)
}

export const filterProducts = (
  productList,
  { productListLength, gender, sort },
) => {
  const totalProductsFilteredByGender = productList.filter(
    product => product.gender === gender,
  )
  const sortedProductsFilteredByGender = [
    ...totalProductsFilteredByGender
      .sort(sortProducts(sort))
      .slice(0, productListLength),
  ]
  return {
    products: [...sortedProductsFilteredByGender],
    isLastPage:
      sortedProductsFilteredByGender.length ===
      totalProductsFilteredByGender.length,
  }
}

export const injectIsWishlistedProp = productList => {
  const wishlistProducts =
    JSON.parse(localStorage.getItem('wishlistProducts')) || []
  return productList.map(
    product =>
      wishlistProducts.includes(product.id)
        ? { ...product, isWishlisted: true }
        : { ...product, isWishlisted: false },
  )
}

export const productSetList = ({ products, isLastPage }) => ({
  type: PRODUCT__SET_LIST,
  payload: { products, isLastPage },
})

export const productSetWishlistProp = ({ id, isWishlisted }) => ({
  type: PRODUCT__SET_WISHLIST_PROP,
  payload: { id, isWishlisted },
})

export const productRefreshList = () => (dispatch, getState) => {
  const { filter } = getState()
  const { products: sortedProducts, isLastPage } = filterProducts(
    productListData,
    filter,
  )
  const productsWithWishlistProp = injectIsWishlistedProp(sortedProducts)

  dispatch(productSetList({ products: productsWithWishlistProp, isLastPage }))
}
