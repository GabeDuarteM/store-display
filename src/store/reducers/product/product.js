import { PRODUCT__SET_LIST, PRODUCT__SET_WISHLIST_PROP } from './productActions'

const product = (state = { products: [], isLastPage: null }, action = {}) => {
  switch (action.type) {
    case PRODUCT__SET_LIST:
      return {
        ...state,
        ...action.payload,
      }
    case PRODUCT__SET_WISHLIST_PROP:
      return {
        ...state,
        products: [
          ...state.products.map(
            elem =>
              elem.id !== action.payload.id
                ? elem
                : { ...elem, isWishlisted: action.payload.isWishlisted },
          ),
        ],
      }

    default:
      return state
  }
}

export default product
