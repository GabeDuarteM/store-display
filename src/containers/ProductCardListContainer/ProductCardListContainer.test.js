import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import ProductCardListContainerConnected, {
  ProductCardListContainer,
} from './ProductCardListContainer'
import ProductCardList from '../../components/ProductCardList'
import { filterLoadMoreProducts as filterLoadMoreProductsAction } from '../../store/reducers/filter/filterActions'
import { productSetWishlistProp } from '../../store/reducers/product/productActions'

const createWrapper = (func, optionalProps) =>
  func(
    <ProductCardListContainer
      filter={{ gender: 'man', productListLength: 18, sort: 'price-desc' }}
      products={[
        {
          id: '5a518050e421e659c6e24c46',
          imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1175/1060f.png',
          gender: 'man',
          rating: 10,
          price: 282.1,
          isNew: true,
          name: 'Perkinsjames Clinegalloway',
          brand: 'VERO MODA',
        },
        {
          id: '5a51805044d1030570bcab76',
          imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1025/1436f.png',
          gender: 'woman',
          rating: 4,
          price: 296.55,
          isNew: false,
          name: 'Henriettagillespie Dellavargas',
          brand: 'NIKE',
        },
      ]}
      handleClickLoadMore={() => {}}
      productRefreshList={() => {}}
      productSetWishlistProp={() => {}}
      {...optionalProps}
    />,
  )

const createWrapperShallow = optionalProps =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = optionalProps => createWrapper(mount, optionalProps)

describe('COMPONENT: <ProductCardListContainer />', () => {
  it('should render without crashing', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it('should have a ProductCardList component', () => {
    const wrapper = createWrapperShallow()
    const hasProductCardList = wrapper.find(ProductCardList).exists()

    expect(hasProductCardList).toBe(true)
  })

  it('should render ProductCardList with the correct props', () => {
    const handleClickLoadMore = jest.fn()
    const toggleIsWishlisted = jest.fn()

    const products = [
      {
        id: '5a518050e421e659c6e24c46',
        imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1175/1060f.png',
        gender: 'man',
        rating: 10,
        price: 282.1,
        isNew: true,
        name: 'Perkinsjames Clinegalloway',
        brand: 'VERO MODA',
      },
      {
        id: '5a51805044d1030570bcab76',
        imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1025/1436f.png',
        gender: 'woman',
        rating: 4,
        price: 296.55,
        isNew: false,
        name: 'Henriettagillespie Dellavargas',
        brand: 'NIKE',
      },
    ]
    const className = 'custom-classname'
    const isLastPage = true

    const wrapper = createWrapperShallow({
      handleClickLoadMore,
      toggleIsWishlisted,
      products,
      className,
      isLastPage,
    })
    const handleClickLoadMoreProp = wrapper.dive().prop('handleClickLoadMore')
    const productsProp = wrapper.dive().prop('products')
    const classNameProp = wrapper.dive().prop('className')
    const isLastPageProp = wrapper.dive().prop('isLastPage')

    expect(handleClickLoadMoreProp).toEqual(handleClickLoadMore)
    expect(productsProp).toEqual(products)
    expect(classNameProp).toEqual(className)
    expect(isLastPageProp).toEqual(isLastPage)
  })

  it('should have a connected component rendering the props correctly', () => {
    const filter = {
      gender: 'man',
      productListLength: 9,
      sort: 'price-asc',
    }
    const store = {
      getState: () => ({
        filter,
        product: {
          products: [],
          isLastPage: true,
        },
      }),
      dispatch: jest.fn(),
      subscribe: () => {},
    }

    const component = shallow(
      <ProductCardListContainerConnected store={store} />,
    ).find(ProductCardListContainer)
    const productsProp = component.prop('products')
    const handleClickLoadMoreProp = component.prop('handleClickLoadMore')
    const productRefreshListProp = component.prop('productRefreshList')
    const productSetWishlistPropProp = component.prop('productSetWishlistProp')

    expect(productsProp).toBeDefined()
    expect(handleClickLoadMoreProp).toBeDefined()
    expect(productRefreshListProp).toBeDefined()
    expect(productSetWishlistPropProp).toBeDefined()

    handleClickLoadMoreProp()
    expect(store.dispatch).toHaveBeenCalled()
    expect(store.dispatch).toHaveBeenCalledWith(filterLoadMoreProductsAction())

    store.dispatch.mockClear()
    productRefreshListProp()
    expect(store.dispatch).toHaveBeenCalled()

    store.dispatch.mockClear()
    const wishlistInfo = { id: 'some id', isWishlisted: true }
    productSetWishlistPropProp(wishlistInfo)
    expect(store.dispatch).toHaveBeenCalled()
    expect(store.dispatch).toHaveBeenCalledWith(
      productSetWishlistProp(wishlistInfo),
    )
  })

  it('should toggle the isWishlisted prop of the product when toggleIsWishlisted is called passing the product id', () => {
    const productSetWishlistPropMock = jest.fn()
    const productId = '5a518050e421e659c6e24c46'
    const wrapper = createWrapperShallow({
      productSetWishlistProp: productSetWishlistPropMock,
    })
    const { toggleIsWishlisted } = wrapper.instance()
    localStorage.clear()

    toggleIsWishlisted(productId)

    let expected = [productId]
    let actual = JSON.parse(localStorage.getItem('wishlistProducts'))

    expect(actual).toEqual(expected)
    expect(productSetWishlistPropMock).toHaveBeenCalledTimes(1)
    expect(productSetWishlistPropMock).toHaveBeenLastCalledWith({
      id: productId,
      isWishlisted: true,
    })

    toggleIsWishlisted(productId)

    expected = []
    actual = JSON.parse(localStorage.getItem('wishlistProducts'))

    expect(actual).toEqual(expected)
    expect(productSetWishlistPropMock).toHaveBeenCalledTimes(2)
    expect(productSetWishlistPropMock).toHaveBeenLastCalledWith({
      id: productId,
      isWishlisted: false,
    })
  })
})
