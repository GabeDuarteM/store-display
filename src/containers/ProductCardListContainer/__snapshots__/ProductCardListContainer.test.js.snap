// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <ProductCardListContainer /> should match component's snapshot 1`] = `
<ProductCardList
  className={null}
  classes={
    Object {
      "loadMoreButton": "ProductCardList-loadMoreButton",
      "loadMoreButtonRoot": "ProductCardList-loadMoreButtonRoot",
      "productCard": "ProductCardList-productCard",
      "productCardList": "ProductCardList-productCardList",
    }
  }
  handleClickLoadMore={[Function]}
  isLastPage={null}
  products={
    Array [
      Object {
        "brand": "VERO MODA",
        "gender": "man",
        "id": "5a518050e421e659c6e24c46",
        "imgUrl": "https://tru-spec-cdn.azureedge.net/media/1175/1060f.png",
        "isNew": true,
        "name": "Perkinsjames Clinegalloway",
        "price": 282.1,
        "rating": 10,
      },
      Object {
        "brand": "NIKE",
        "gender": "woman",
        "id": "5a51805044d1030570bcab76",
        "imgUrl": "https://tru-spec-cdn.azureedge.net/media/1025/1436f.png",
        "isNew": false,
        "name": "Henriettagillespie Dellavargas",
        "price": 296.55,
        "rating": 4,
      },
    ]
  }
  toggleIsWishlisted={[Function]}
/>
`;
